hw_diag/utilities/dbus_proxies.py

```diff
@@ -0,0 +1,221 @@
+
+    DBUS_PROPERTIES_IF = 'org.freedesktop.DBus.Properties'
+    DBUS_OBJECTMANAGER_IF = 'org.freedesktop.DBus.ObjectManager'
+
+    DBUS_MM1_SERVICE = 'org.freedesktop.ModemManager1'
+    DBUS_MM1_PATH = '/org/freedesktop/ModemManager1'
+    DBUS_MM1_IF = 'org.freedesktop.ModemManager1'
+    DBUS_MM1_MODEM_IF = 'org.freedesktop.ModemManager1.Modem'
+
+    SYSTEMD_SERVICE = 'org.freedesktop.systemd1'
+    SYSTEMD_PATH = '/org/freedesktop/systemd1'
+    SYSTEMD_MANAGER_IF = 'org.freedesktop.systemd1.Manager'
+    SYSTEMD_UNIT_IF  = 'org.freedesktop.systemd1.Unit'
+
+class DBusObject(object):
+    '''Base class for various dbus objects'''
+
+    def __init__(self, service, path, interface):
+        self._system_bus = dbus.SystemBus()
+        self._object_proxy = self._system_bus.get_object(service, path)
+        self._object_if = dbus.Interface(self._object_proxy, interface)
+        self._properties_iface = dbus.Interface(self._object_proxy, DBusIds.DBUS_PROPERTIES_IF)
+
+    def get_properties(self):
+        props = self._properties_iface.GetAll(self._object_if.dbus_interface)
+        return props
+
+    def get_property(self, property_name):
+        prop = self._properties_iface.Get(self._object_if.dbus_interface, property_name)
+        return prop
+
+class Modem(DBusObject):
+    '''
+    Encapsulates modem manager's modem dbus object and provides a few more
+    convenience methods.
+    Note:: AT commands assume quectel modem
+    '''
+
+    FIRMWARE_VER_AT_CMD  = 'AT+QGMR'
+    UE_MODE_SETTING_WRITE_AT_CMD = 'AT+QNVFW="/nv/item_files/modem/mmode/ue_usage_setting"'
+    UE_MODE_SETTING_READ_AT_CMD = 'AT+QNVFR="/nv/item_files/modem/mmode/ue_usage_setting"'
+    UE_MODE_DATA_ONLY_VALUE = '01'
+    SERVICE_DOMAIN_AT_CMD = 'AT+QCFG=“servicedomain”'
+    SERVICE_DOMAIN_PS_VALUE = '1,1'
+
+    def __init__(self, modem_obj_path):
+        super(Modem, self).__init__(DBusIds.DBUS_MM1_SERVICE,
+                                    modem_obj_path,
+                                    DBusIds.DBUS_MM1_MODEM_IF)
+
+    def at_command(self, cmd, timeout=2000):
+        response = self._object_if.Command(cmd, timeout)
+        return response
+
+    def get_fw_version(self):
+        return self.at_command(self.FIRMWARE_VER_AT_CMD)
+
+    def get_UE_mode(self):
+        response = self.at_command(self.UE_MODE_SETTING_READ_AT_CMD)
+        values = response.strip().split(' ')
+        if len(values) == 2:
+            return values[1]
+
+    def set_UE_Mode(self, value = UE_MODE_DATA_ONLY_VALUE):
+        return self.at_command(self.UE_MODE_SETTING_WRITE_AT_CMD + ',' + value)
+
+    def get_service_domain(self, value = SERVICE_DOMAIN_PS_VALUE):
+        return self.at_command(self.SERVICE_DOMAIN_AT_CMD)
+
+    def set_service_domain(self, value = SERVICE_DOMAIN_PS_VALUE):
+        return self.at_command(self.SERVICE_DOMAIN_AT_CMD + ',' + value)
+
+    def ensure_data_centric_mode(self):
+        if self.get_UE_mode() != self.UE_MODE_DATA_ONLY_VALUE:
```

Comment:
Could you justify this decision one more time? How will this work when new modules are shipped with newer firmware?

Reply:
this should work with all firmware versions. It is recommended to set data centric if there is no use of voice/sms irrespective of firmware version.